server/src/main/scala/org/http4s/server/middleware/authentication/package.scala

```diff
@@ -7,9 +7,17 @@ import scalaz.concurrent.Task
 import org.http4s.headers._
 
 package object authentication {
+  type Realm = String
+  type User = String
+  type ClientPasssword = String
   // A function mapping (realm, username) to password, None if no password
   // exists for that (realm, username) pair.
-  type AuthenticationStore = (String, String) => Task[Option[String]]
+  type AuthenticationStore = (Realm, User) => Task[Option[String]]
+  /**
+    *  Validates a plaintext password (presumably by comparing it to a hashed value).
+    *  Some(true) means the password given is valid.
```

Comment:
Maybe a sealed trait is more expressive for this ternary logic: Authenticated | IncorrectPassword | NoSuchUser.

It's not uncommon to have authentication backends that don't disclose whether the account exists on failure.  Does this require a fourth result type to express "It's not valid, but I'm not going to tell you why?"  I'm not sure.